File: NekrasovskyAPP/NekrasovskyAPP/Services/SignalRService.cs
using Microsoft.AspNetCore.SignalR.Client;
using System.Text.Json;
using Microsoft.Extensions.Logging;

namespace NekrasovskyAPP.Services
{
    public class SignalRService : ISignalRService, IDisposable
    {
        private HubConnection? _connection;
        private readonly ILogger<SignalRService>? _logger;
        private readonly string _baseUrl;
        private event Action<string, string, string>? OnAlarmNotificationReceived;

        public HubConnection? Connection => _connection;
        public bool IsConnected => _connection?.State == HubConnectionState.Connected;
        
        public void SetOnAlarmNotification(Action<string, string, string> callback)
        {
            OnAlarmNotificationReceived += callback;
        }

        public SignalRService(ILogger<SignalRService>? logger = null)
        {
            _logger = logger;
            _baseUrl = GetBaseUrl();
        }

        private static string GetBaseUrl()
        {
#if ANDROID
            return "http://192.168.1.121:9000/";
#else
            return "http://localhost:9000/";
#endif
        }

        public async Task ConnectAsync()
        {
            if (_connection != null && IsConnected)
            {
                return;
            }

            try
            {
                _connection = new HubConnectionBuilder()
                    .WithUrl($"{_baseUrl}hubs/notifications")
                    .WithAutomaticReconnect()
                    .Build();

                // Обработка переподключения
                _connection.Reconnecting += error =>
                {
                    _logger?.LogWarning("SignalR переподключение... {Error}", error?.Message);
                    return Task.CompletedTask;
                };

                _connection.Reconnected += connectionId =>
                {
                    _logger?.LogInformation("SignalR переподключен. ConnectionId: {ConnectionId}", connectionId);
                    // Автоматически подписываемся на уведомления при переподключении
                    _ = SubscribeToAlarmNotificationsAsync();
                    return Task.CompletedTask;
                };

                _connection.Closed += error =>
                {
                    _logger?.LogError("SignalR соединение закрыто. {Error}", error?.Message);
                    return Task.CompletedTask;
                };

                // Настраиваем обработчик уведомлений о тревогах
                _connection.On<object>("AlarmNotification", (notification) =>
                {
                    try
                    {
                        // Парсим уведомление
                        var json = System.Text.Json.JsonSerializer.Serialize(notification);
                        var doc = System.Text.Json.JsonDocument.Parse(json);
                        var root = doc.RootElement;

                        var message = root.TryGetProperty("message", out var msgProp) 
                            ? msgProp.GetString() ?? "СОБЫТИЕ ТРЕВОГИ!" 
                            : "СОБЫТИЕ ТРЕВОГИ!";
                        var location = root.TryGetProperty("location", out var locProp) 
                            ? locProp.GetString() ?? "Неизвестно" 
                            : "Неизвестно";
                        var user = root.TryGetProperty("user", out var userProp) 
                            ? userProp.GetString() ?? "Неизвестный пользователь" 
                            : "Неизвестный пользователь";

                        OnAlarmNotificationReceived?.Invoke(message, location, user);
                    }
                    catch (Exception ex)
                    {
                        _logger?.LogError(ex, "Ошибка обработки уведомления о тревоге");
                    }
                });

                await _connection.StartAsync();
                _logger?.LogInformation("SignalR подключен успешно");

                // Подписываемся на уведомления о тревогах
                await SubscribeToAlarmNotificationsAsync();
            }
            catch (Exception ex)
            {
                _logger?.LogError(ex, "Ошибка подключения к SignalR Hub");
                throw;
            }
        }

        public async Task DisconnectAsync()
        {
            if (_connection != null)
            {
                try
                {
                    await _connection.StopAsync();
                    await _connection.DisposeAsync();
                    _connection = null;
                    _logger?.LogInformation("SignalR отключен");
                }
                catch (Exception ex)
                {
                    _logger?.LogError(ex, "Ошибка при отключении от SignalR");
                }
            }
        }

        public async Task SubscribeToAlarmNotificationsAsync()
        {
            if (_connection != null && IsConnected)
            {
                try
                {
                    await _connection.InvokeAsync("SubscribeToAlarmNotifications");
                    _logger?.LogInformation("Подписка на уведомления о тревогах активирована");
                }
                catch (Exception ex)
                {
                    _logger?.LogError(ex, "Ошибка при подписке на уведомления о тревогах");
                }
            }
        }

        public void Dispose()
        {
            DisconnectAsync().Wait(5000);
        }
    }
}

File: android-app/app/src/main/java/com/nekrasovsky/android/data/signalr/SignalRService.kt
package com.nekrasovsky.android.data.signalr

import android.util.Log
import com.microsoft.signalr.HubConnection
import com.microsoft.signalr.HubConnectionBuilder
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONObject

class SignalRService {
    private var hubConnection: HubConnection? = null
    private val baseUrl = "http://192.168.1.121:9000/" // TODO: Вынести в конфигурацию
    private var onAlarmNotification: ((String, String, String) -> Unit)? = null

    fun setOnAlarmNotification(callback: (message: String, location: String, user: String) -> Unit) {
        this.onAlarmNotification = callback
    }

    suspend fun connect() {
        try {
            if (hubConnection?.connectionState == com.microsoft.signalr.HubConnectionState.CONNECTED) {
                return
            }

            hubConnection = HubConnectionBuilder.create("${baseUrl}hubs/notifications")
                .withAutomaticReconnect()
                .build()

            // Обработка уведомлений о тревогах
            hubConnection?.on("AlarmNotification") { notification ->
                try {
                    val json = JSONObject(notification.toString())
                    val message = json.optString("message", "СОБЫТИЕ ТРЕВОГИ!")
                    val location = json.optString("location", "Неизвестно")
                    val user = json.optString("user", "Неизвестный пользователь")
                    val sound = json.optBoolean("sound", true)

                    onAlarmNotification?.invoke(message, location, user)
                } catch (e: Exception) {
                    Log.e("SignalRService", "Ошибка обработки уведомления", e)
                }
            }

            hubConnection?.start()
            Log.d("SignalRService", "SignalR подключен")

            // Подписываемся на уведомления о тревогах
            subscribeToAlarmNotifications()
        } catch (e: Exception) {
            Log.e("SignalRService", "Ошибка подключения к SignalR", e)
        }
    }

    suspend fun disconnect() {
        try {
            hubConnection?.stop()
            hubConnection = null
            Log.d("SignalRService", "SignalR отключен")
        } catch (e: Exception) {
            Log.e("SignalRService", "Ошибка отключения от SignalR", e)
        }
    }

    private suspend fun subscribeToAlarmNotifications() {
        try {
            hubConnection?.invoke("SubscribeToAlarmNotifications")
            Log.d("SignalRService", "Подписка на уведомления о тревогах активирована")
        } catch (e: Exception) {
            Log.e("SignalRService", "Ошибка подписки на уведомления", e)
        }
    }

    fun isConnected(): Boolean {
        return hubConnection?.connectionState == com.microsoft.signalr.HubConnectionState.CONNECTED
    }
}
